<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jelmer Poelstra">

<title>Software at OSC</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">RNAseq analysis</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html" rel="" target="">
 <span class="menu-text"><i class="fa-solid fa-home" aria-label="home"></i></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../modules/01_intro.html" rel="" target="">
 <span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-computing-skills" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">1. Computing skills</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-computing-skills">    
        <li>
    <a class="dropdown-item" href="../modules/02_osc.html" rel="" target="">
 <span class="dropdown-text">- OSC</span></a>
  </li>  
        <li class="dropdown-header">- VS Code [TBA]</li>
        <li class="dropdown-header">- The Unix shell [TBA]</li>
        <li class="dropdown-header">- Shell scripting [TBA]</li>
        <li class="dropdown-header">- Software at OSC [TBA]</li>
        <li class="dropdown-header">- SLURM compute jobs [TBA]</li>
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-from-reads-to-counts" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">2. From reads to counts</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-from-reads-to-counts">    
        <li class="dropdown-header">- Read QC with <code>FastQC</code> and <code>MultiQC</code> [TBA]</li>
        <li class="dropdown-header">- Trimming with <code>TrimGalore</code> [TBA]</li>
        <li class="dropdown-header">- rRNA removal with <code>SortMeRNA</code> [TBA]</li>
        <li class="dropdown-header">- Alignment with <code>STAR</code> [TBA]</li>
        <li class="dropdown-header">- Counting with <code>Salmon</code> [TBA]</li>
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-analyzing-the-counts" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">3. Analyzing the counts</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-analyzing-the-counts">    
        <li class="dropdown-header">- Creating a DESeq2 object [TBA]</li>
        <li class="dropdown-header">- PCA [TBA]</li>
        <li class="dropdown-header">- Differential expression analysis [TBA]</li>
        <li class="dropdown-header">- GO enrichment analysis [TBA]</li>
        <li class="dropdown-header">- KEGG enrichment analysis [TBA]</li>
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-reference-material" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Reference material</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-reference-material">    
        <li class="dropdown-header">Specific tutorials</li>
        <li>
    <a class="dropdown-item" href="../info/osc_transfer.html" rel="" target="">
 <span class="dropdown-text">- File transfer to and from OSC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../info/osc_ssh.html" rel="" target="">
 <span class="dropdown-text">- Using OSC with SSH</span></a>
  </li>  
        <li class="dropdown-header">General info</li>
        <li>
    <a class="dropdown-item" href="../info/resources.html" rel="" target="">
 <span class="dropdown-text">- Further Resources</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../info/glossary.html" rel="" target="">
 <span class="dropdown-text">- Glossary</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../info/about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/mcic-osu/rnaseq-intro" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#setup" id="toc-setup" class="nav-link active" data-scroll-target="#setup"><span class="header-section-number">1</span> Setup</a></li>
  <li><a href="#running-command-line-programs" id="toc-running-command-line-programs" class="nav-link" data-scroll-target="#running-command-line-programs"><span class="header-section-number">2</span> Running command-line programs</a></li>
  <li><a href="#software-at-osc-with-lmod" id="toc-software-at-osc-with-lmod" class="nav-link" data-scroll-target="#software-at-osc-with-lmod"><span class="header-section-number">3</span> Software at OSC with <code>Lmod</code></a>
  <ul class="collapse">
  <li><a href="#checking-for-available-software" id="toc-checking-for-available-software" class="nav-link" data-scroll-target="#checking-for-available-software"><span class="header-section-number">3.1</span> Checking for available software</a></li>
  <li><a href="#loading-software" id="toc-loading-software" class="nav-link" data-scroll-target="#loading-software"><span class="header-section-number">3.2</span> Loading software</a></li>
  <li><a href="#a-practical-example" id="toc-a-practical-example" class="nav-link" data-scroll-target="#a-practical-example"><span class="header-section-number">3.3</span> A practical example</a></li>
  </ul></li>
  <li><a href="#when-software-isnt-installed-at-osc" id="toc-when-software-isnt-installed-at-osc" class="nav-link" data-scroll-target="#when-software-isnt-installed-at-osc"><span class="header-section-number">4</span> When software isn’t installed at OSC</a></li>
  <li><a href="#using-conda" id="toc-using-conda" class="nav-link" data-scroll-target="#using-conda"><span class="header-section-number">5</span> Using conda</a>
  <ul class="collapse">
  <li><a href="#loading-the-miniconda-module" id="toc-loading-the-miniconda-module" class="nav-link" data-scroll-target="#loading-the-miniconda-module"><span class="header-section-number">5.1</span> Loading the (mini)conda module</a></li>
  <li><a href="#one-time-conda-configuration" id="toc-one-time-conda-configuration" class="nav-link" data-scroll-target="#one-time-conda-configuration"><span class="header-section-number">5.2</span> One-time conda configuration</a></li>
  <li><a href="#example-creating-an-environment-for-cutadapt" id="toc-example-creating-an-environment-for-cutadapt" class="nav-link" data-scroll-target="#example-creating-an-environment-for-cutadapt"><span class="header-section-number">5.3</span> Example: Creating an environment for <code>cutadapt</code></a></li>
  <li><a href="#activating-conda-environments" id="toc-activating-conda-environments" class="nav-link" data-scroll-target="#activating-conda-environments"><span class="header-section-number">5.4</span> Activating conda environments</a></li>
  <li><a href="#creating-an-environment-for-any-program" id="toc-creating-an-environment-for-any-program" class="nav-link" data-scroll-target="#creating-an-environment-for-any-program"><span class="header-section-number">5.5</span> Creating an environment for <em>any</em> program</a></li>
  <li><a href="#lines-to-add-to-your-bash-script" id="toc-lines-to-add-to-your-bash-script" class="nav-link" data-scroll-target="#lines-to-add-to-your-bash-script"><span class="header-section-number">5.6</span> Lines to add to your Bash script</a></li>
  </ul></li>
  <li><a href="#addendum-a-few-other-useful-conda-commands" id="toc-addendum-a-few-other-useful-conda-commands" class="nav-link" data-scroll-target="#addendum-a-few-other-useful-conda-commands"><span class="header-section-number">6</span> Addendum: a few other useful conda commands</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/mcic-osu/rnaseq-intro/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Using Software at OSC</h1>
<p class="subtitle lead">Loading existing modules and installing and using software with conda</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Jelmer Poelstra </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<hr>
<p>So far, we have only used commands that are available in any Unix shell. But to actually analyze genomics data sets, we also need to use specialized bioinformatics software.</p>
<p>Most software that is already installed at OSC must nevertheless be “loaded” (“activated”) before we can use it — and if our software of choice is not installed, we have to do so ourselves. We will cover those topics in this module.</p>
<section id="setup" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="setup"><span class="header-section-number">1</span> Setup</h2>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Starting a VS Code session with an active terminal (click here)
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ol type="1">
<li><p>Log in to OSC at <a href="https://ondemand.osc.edu" class="uri">https://ondemand.osc.edu</a>.</p></li>
<li><p>In the blue top bar, select <code>Interactive Apps</code> and then <code>Code Server</code>.</p></li>
<li><p>In the form that appears:</p>
<ul>
<li>Enter <code>4</code> or more in the box <code>Number of hours</code></li>
<li><strong>To avoid having to switch folders within VS Code</strong>, enter <code>/fs/ess/scratch/PAS2250/participants/&lt;your-folder&gt;</code> in the box <code>Working Directory</code> (replace <code>&lt;your-folder&gt;</code> by the actual name of your folder).</li>
<li>Click <code>Launch</code>.</li>
</ul></li>
<li><p>On the next page, once the top bar of the box is green and says <code>Runnning</code>, click <code>Connect to VS Code</code>.</p></li>
<li><p>Open a terminal: <i class="fa-solid fa-bars" aria-label="bars"></i> &nbsp; =&gt; <code>Terminal</code> =&gt; <code>New Terminal</code>.</p></li>
<li><p>In the terminal, type <code>bash</code> and press <kbd>Enter</kbd>.</p></li>
<li><p>Type <code>pwd</code> in the termain to check you are in <code>/fs/ess/scratch/PAS2250</code>.</p>
<p>If not, click <i class="fa-solid fa-bars" aria-label="bars"></i> &nbsp; =&gt; &nbsp; <code>File</code> &nbsp; =&gt; &nbsp; <code>Open Folder</code> and enter <code>/fs/ess/scratch/PAS2250/&lt;your-folder&gt;</code>.</p></li>
</ol>
</div>
</div>
</div>
<p><br></p>
</section>
<section id="running-command-line-programs" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="running-command-line-programs"><span class="header-section-number">2</span> Running command-line programs</h2>
<p>As pointed out in the introduction to the workshop, bioinformatics software (programs) that we use to analyze genomic data are typically run from the command line. That is, they have “command-line interfaces” (<em>CLI</em>s) rather than “graphical user interfaces” (<em>GUI</em>s), and are run using commands that are structurally very similar to how we’ve been using basic Unix commands.</p>
<p>For instance, we can run the program <code>FastQC</code> as follows, instructing it to process the <code>FASTQ</code> file <code>sampleA.fastq.gz</code> with default options:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> sampleA.fastq.gz       <span class="co"># Don't run</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>So, with all the scaffolding we have learned in the previous modules, we only need to make small modifications to have our scripts run command-line programs. But, we first need to load and/or install these programs.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Running inside a script or interactively
</div>
</div>
<div class="callout-body-container callout-body">
<p>Like any other command, we could in principle run the line of code above either in our interactive shell or from inside a script. In practice, it is <strong>better to do this in a script</strong>, especially at OSC, because:</p>
<ol type="1">
<li>Such programs typically take a while to run</li>
<li>We are not supposed to run processes that use significant resources on login nodes</li>
<li>We can run the same script simultaneously for different input files.</li>
</ol>
</div>
</div>
<p><br></p>
</section>
<section id="software-at-osc-with-lmod" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="software-at-osc-with-lmod"><span class="header-section-number">3</span> Software at OSC with <code>Lmod</code></h2>
<p>OSC administrators manage software with the <code>Lmod</code> system of software modules. For us users, this means that even though a lot of software is installed, <strong>most of it can only be used after we explicitly load it.</strong><br>
(That may seem like a drag, but on the upside, this practice enables the use of different versions of the same software, and of mutually incompatible software on a single system.)</p>
<section id="checking-for-available-software" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="checking-for-available-software"><span class="header-section-number">3.1</span> Checking for available software</h3>
<p>The OSC website has a <a href="https://www.osc.edu/resources/available_software/software_list">list of software that has been installed at OSC</a>. You can also search for available software in the shell using two subtly different commands:</p>
<ul>
<li><code>module spider</code> lists modules that are installed.</li>
<li><code>module avail</code> lists modules that <em>can be directly loaded</em>, given the current environment (i.e., depending on which other software has been loaded).</li>
</ul>
<p>Simply running <code>module spider</code> or <code>module avail</code> would spit out complete lists — more usefully, we can add <strong>search terms</strong> as arguments to these commands:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> spider python</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="bash-out">
<table class="table">
<tbody>
<tr class="odd">
<td style="text-align: left;">python:</td>
</tr>
</tbody>
</table>
<pre><code> Versions:
    python/2.7-conda5.2
    python/3.6-conda5.2
    python/3.7-2019.10</code></pre>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> avail python</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="bash-out">
<p>python/2.7-conda5.2 &nbsp; &nbsp; &nbsp; &nbsp; python/3.6-conda5.2 (D) &nbsp; &nbsp; &nbsp; &nbsp; python/3.7-2019.10</p>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<code>(D)</code> = default version
</div>
</div>
<div class="callout-body-container callout-body">
<p>The <code>(D)</code> in the output above marks the default version of the program; that is, the version of the program that would be loaded if we don’t specify a version (see examples below).</p>
</div>
</div>
<p><br></p>
</section>
<section id="loading-software" class="level3" data-number="3.2">
<h3 data-number="3.2" class="anchored" data-anchor-id="loading-software"><span class="header-section-number">3.2</span> Loading software</h3>
<p>All other <code>Lmod</code> software functionality is also accessed using <code>module</code> “subcommands” (we call <code>module</code> the command and e.g.&nbsp;<code>spider</code> the subcommand). For instance, to load a module:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load a module:</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load python              <span class="co"># Load the default version</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load python/3.7-2019.10  <span class="co"># Load a specific version (copy from module spider output)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To check which modules have been loaded (the list includes automatically loaded modules):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> list</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="bash-out">
<p>Currently Loaded Modules:<br>
&nbsp; &nbsp; 1) xalt/latest &nbsp; &nbsp; &nbsp; 2) gcc-compatibility/8.4.0 &nbsp; &nbsp; &nbsp; 3) intel/19.0.5 &nbsp; &nbsp; &nbsp; 4) mvapich2/2.3.3 &nbsp; &nbsp; &nbsp; 5) modules/sp2020</p>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Unloading modules
</div>
</div>
<div class="callout-body-container callout-body">
<p>Occasionally useful when running into conflicting (mutually incompatible) modules:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> unload python        <span class="co"># Unload a module</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> purge                <span class="co"># Unload all modules</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p><br></p>
</section>
<section id="a-practical-example" class="level3" data-number="3.3">
<h3 data-number="3.3" class="anchored" data-anchor-id="a-practical-example"><span class="header-section-number">3.3</span> A practical example</h3>
<p>Let’s load a <em>very</em> commonly used bioinformatics program that we will also use in examples later on: FastQC. FastQC performs quality control (hence: “QC”) on FASTQ files.</p>
<p>First, let’s test that we indeed <em>cannot</em> currently use <code>fastqc</code> by running <code>fastqc</code> with the <code>--help</code> flag:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">--help</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="bash-out">
<p>bash: fastqc: command not found</p>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Help!
</div>
</div>
<div class="callout-body-container callout-body">
<p>A solid majority of command-line programs can be run with with a <code>--help</code> (and/or <code>-h</code>) flag, and this is often a good thing to try first, since it will tell use whether we can use the program, and if we can, we immediately get some usage information.</p>
</div>
</div>
<p>Next, let’s check whether FastQC is available at OSC, and if so, in which versions:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> avail fastqc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="bash-out">
<p>fastqc/0.11.8</p>
</div>
<p>There is only one version available (<code>0.11.8</code>), which means that <code>module load fastqc</code> and <code>module load fastqc/0.11.8</code> would each load that same version.</p>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
What might still be a reason to specify the version when we load FastQC?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>When we use it inside a script:</p>
<ul>
<li><p>This would ensure that when we run the same script a year later, the same version would be used (assuming it hasn’t been removed) — otherwise, it’s possible a newer version would has been installed in the meantime, which might produce different results.</p></li>
<li><p>It will make it easy to see which version we used, which is something we typically report in papers.</p></li>
</ul>
</div>
</div>
</div>
<p>Let’s load the FastQC module:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load fastqc/0.11.8</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>After we have loaded the module, we can retry our <code>--help</code> attempt:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">--help</span> <span class="kw">|</span> <span class="fu">head</span>     <span class="co"># I'm piping into head to avoid pages worth of output </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="bash-out">
<pre><code>        FastQC - A high throughput sequence QC analysis tool</code></pre>
<p>SYNOPSIS</p>
<pre><code>    fastqc seqfile1 seqfile2 .. seqfileN

fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] 
       [-c contaminant file] seqfile1 .. seqfileN  
       </code></pre>
</div>
<p><br></p>
</section>
</section>
<section id="when-software-isnt-installed-at-osc" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="when-software-isnt-installed-at-osc"><span class="header-section-number">4</span> When software isn’t installed at OSC</h2>
<p>It’s not too uncommon that software you need for your project is not installed at OSC, or that you need a more recent version of the software than is available. The main options available to you in such a case are to:</p>
<ul>
<li><p><strong>“Manually” install the software</strong>, which in the best case involves downloading a directly functioning binary (executable), but more commonly requires you to “compile” (build) the program. This is sometimes straightforward but can also become extremely tricky, especially at OSC where you don’t have “administrator privileges”<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> at OSC and will often have difficulties with “dependencies”<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>.</p></li>
<li><p>Send an <strong>email to <a href="mailto:oschelp@osc.edu">OSC Help</a></strong>. They might be able to help you with your installation, or in case of commonly used software, might be willing to perform a system-wide installation (that is, making it available through <code>module</code>). #TODO: Refer to the software form in https://www.osc.edu/resources/getting_started/supercomputing_faq#req</p></li>
<li><p>Use <strong><code>Apptainer</code> / <code>Singularity</code> “containers”</strong>. Containers are self-contained software environments that include operating systems, akin to mini virtual machines.</p></li>
<li><p>Use <strong><code>conda</code></strong>, which creates software environments that are activated like in the <code>module</code> system.</p></li>
</ul>
<p>Conda and containers are useful not only at OSC, where they <strong>bypass issues with dependencies and administrator privileges</strong>, <strong>but more generally, for reproducible and portable software environments</strong>. They also allow you to easily maintain distinct “environments”, each with a different version of the same software, or with mutually incompatible software.</p>
<p>We will teach conda here because it is easier to learn and use than containers, and because nearly all open-source bioinformatics software is available as a conda package.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
When to use containers instead of Conda
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>If you need to use software that requires a different Operating System (OS) or OS version than the one at OSC.</li>
<li>If you want or require even greater reproducibility and portability to create an isolated environment that can be exported and used anywhere.</li>
</ul>
</div>
</div>
<p><br></p>
</section>
<section id="using-conda" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="using-conda"><span class="header-section-number">5</span> Using conda</h2>
<p>Conda creates so-called <strong>environments</strong> in which you can install one or more software packages. As mentioned above, these environments are activated and deactivated in a similar manner as with the <code>Lmod</code> system – but the key difference is that we can create and manage these environments ourselves.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Use one environment per program (as here) or one per research project
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Below are two reasonable ways to organize your conda environments, and their advantages:</p>
<ul>
<li>Have one environment per program (my preference)
<ul>
<li>Easier to keep an overview of what you have installed</li>
<li>No need to reinstall the same program across different projects</li>
<li>Less risk of running into problems with your environment due to mutually incompatible software and complicated dependency situations</li>
</ul></li>
<li>Have one environment per research project
<ul>
<li>You just need to activate that one environment when you’re working on your project.</li>
<li>Easier when you need to share your entire project with someone else (or yourself) on a different (super)computer.</li>
</ul></li>
</ul>
<p>Its <strong>not</strong> recommended to simply install all programs across all projects in one environment. This doesn’t benefit reproducibility and your environment is likely to sooner or later stop functioning properly.</p>
<p>A side note is that even when you want to install a single program, multiple programs <em>are</em> in fact nearly always installed: the programs that your target program depends on (“dependencies”).</p>
</div>
</div>
</div>
<p><br></p>
<section id="loading-the-miniconda-module" class="level3" data-number="5.1">
<h3 data-number="5.1" class="anchored" data-anchor-id="loading-the-miniconda-module"><span class="header-section-number">5.1</span> Loading the (mini)conda module</h3>
<p>While it is also fairly straightforward to install conda for yourself <a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>, we will use OSC’s system-wide installation of conda in this workshop. Therefore, we first need to use a <code>module load</code> command to make it available:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># (The most common installation of conda is actually called "miniconda")</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load miniconda3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
</section>
<section id="one-time-conda-configuration" class="level3" data-number="5.2">
<h3 data-number="5.2" class="anchored" data-anchor-id="one-time-conda-configuration"><span class="header-section-number">5.2</span> One-time conda configuration</h3>
<p>We will also do some one-time configuration, which will set the <strong>conda “channels”</strong> (basically, software repositories) that we want to use when we install software. This config also includes setting relative <em>priorities</em> among channels, since one software package may be available from multiple channels.</p>
<p>Like with <code>module</code> commands, conda commands consist of two parts, the conda command itself and a subcommand, such as <code>config</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> config <span class="at">--add</span> channels defaults     <span class="co"># Added first =&gt; lowest priority</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> config <span class="at">--add</span> channels bioconda</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> config <span class="at">--add</span> channels conda-forge  <span class="co"># Added last =&gt; highest priority</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s check whether this configuration step worked:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> config <span class="at">--get</span> channels</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="bash-out">
<p>–add channels ‘defaults’ &nbsp; &nbsp; &nbsp; # lowest priority<br>
–add channels ‘bioconda’<br>
–add channels ‘conda-forge’ &nbsp; &nbsp; &nbsp; # highest priority</p>
</div>
<p><br></p>
</section>
<section id="example-creating-an-environment-for-cutadapt" class="level3" data-number="5.3">
<h3 data-number="5.3" class="anchored" data-anchor-id="example-creating-an-environment-for-cutadapt"><span class="header-section-number">5.3</span> Example: Creating an environment for <code>cutadapt</code></h3>
<p>To practice using conda, we will now create a conda environment with the program <code>cutadapt</code> installed.</p>
<p><code>cutadapt</code> is a commonly used program to remove adapters or primers from sequence reads in <code>FASTQ</code> files — in particular, it is ubiquitous for primer removal in (e.g.&nbsp;16S rRNA) microbiome metabarcoding studies. But there is no <code>Lmod</code> module on OSC for it, so if we want to use it, our best option is to resort to conda.</p>
<p>Here is the command to create a new environment and install <code>cutadapt</code> into that environment:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-y</span> <span class="at">-n</span> cutadapt <span class="at">-c</span> bioconda cutadapt   <span class="co"># Don't run this</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s break the above command down:</p>
<ul>
<li><p><strong><code>create</code></strong> is the conda subcommand to create a new environment.</p></li>
<li><p><strong><code>-y</code></strong> is a flag that prevents us from being asked to confirm installation.</p></li>
<li><p>Following the <strong><code>-n</code></strong> option, we can specify the name of the environment, so <strong><code>-n cutadapt</code></strong> means that we want our environment to be called <code>cutadapt</code>. We can use whatever name we like for the environment, but of course a descriptive yet concise name is a good idea. Since we are making a single-program environment, it makes sense to simply name it after the program.</p></li>
<li><p>Following the <strong><code>-c</code></strong> option, we can specify a channel from which we want to install, so <strong><code>-c bioconda</code></strong> indicates we want to use the <code>bioconda</code> channel. (Given that we’ve done some config above, this is not always necessary, but it can be good to be explicit.)</p></li>
<li><p>The <strong><code>cutadapt</code></strong> at the end of the line simply tells conda to install the package of that name. This is a “<em>positional</em>” argument to the command (note that there’s no option like <code>-s</code> before it): we put any software package(s) we want to install at the end of the command.</p></li>
</ul>
<section id="specifying-a-version" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="specifying-a-version">Specifying a version</h4>
<p>If we want to be explicit about the version we want to install, we can add the version after <code>=</code> following the package name. We do that below, and we also include the version in the environment name.</p>
<p><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> <strong><em>Let’s run the command above and see if we can install <code>cutadapt</code></em></strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-y</span> <span class="at">-n</span> cutadapt-4.1 <span class="at">-c</span> bioconda cutadapt=4.1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="bash-out">
<p>Collecting package metadata (current_repodata.json): done<br>
Solving environment:</p>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
See the full output when I ran this command
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><br></p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre class="bash-out"><code>Collecting package metadata (current_repodata.json): done
Solving environment: done


==&gt; WARNING: A newer version of conda exists. &lt;==
  current version: 4.10.3
  latest version: 4.13.0

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /fs/project/PAS0471/jelmer/conda/cutadapt-TMP

  added / updated specs:
    - cutadapt=4.1


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    cutadapt-4.1               |  py310h1425a21_1         211 KB  bioconda
    dnaio-0.9.1                |  py310h1425a21_1          80 KB  bioconda
    libsqlite-3.39.2           |       h753d276_1         789 KB  conda-forge
    openssl-3.0.5              |       h166bdaf_1         2.8 MB  conda-forge
    pip-22.2.2                 |     pyhd8ed1ab_0         1.5 MB  conda-forge
    python-isal-1.0.1          |  py310h5764c6d_0          47 KB  conda-forge
    setuptools-65.0.2          |  py310hff52083_0         1.4 MB  conda-forge
    sqlite-3.39.2              |       h4ff8645_1         788 KB  conda-forge
    tzdata-2022c               |       h191b570_0         119 KB  conda-forge
    xopen-1.6.0                |  py310hff52083_0          27 KB  conda-forge
    xz-5.2.6                   |       h166bdaf_0         409 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         8.1 MB

The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-2_gnu
  bzip2              conda-forge/linux-64::bzip2-1.0.8-h7f98852_4
  ca-certificates    conda-forge/linux-64::ca-certificates-2022.6.15-ha878542_0
  cutadapt           bioconda/linux-64::cutadapt-4.1-py310h1425a21_1
  dnaio              bioconda/linux-64::dnaio-0.9.1-py310h1425a21_1
  isa-l              conda-forge/linux-64::isa-l-2.30.0-ha770c72_4
  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.36.1-hea4e1c9_2
  libffi             conda-forge/linux-64::libffi-3.4.2-h7f98852_5
  libgcc-ng          conda-forge/linux-64::libgcc-ng-12.1.0-h8d9b700_16
  libgomp            conda-forge/linux-64::libgomp-12.1.0-h8d9b700_16
  libnsl             conda-forge/linux-64::libnsl-2.0.0-h7f98852_0
  libsqlite          conda-forge/linux-64::libsqlite-3.39.2-h753d276_1
  libuuid            conda-forge/linux-64::libuuid-2.32.1-h7f98852_1000
  libzlib            conda-forge/linux-64::libzlib-1.2.12-h166bdaf_2
  ncurses            conda-forge/linux-64::ncurses-6.3-h27087fc_1
  openssl            conda-forge/linux-64::openssl-3.0.5-h166bdaf_1
  pbzip2             conda-forge/linux-64::pbzip2-1.1.13-0
  pigz               conda-forge/linux-64::pigz-2.6-h27826a3_0
  pip                conda-forge/noarch::pip-22.2.2-pyhd8ed1ab_0
  python             conda-forge/linux-64::python-3.10.5-ha86cf86_0_cpython
  python-isal        conda-forge/linux-64::python-isal-1.0.1-py310h5764c6d_0
  python_abi         conda-forge/linux-64::python_abi-3.10-2_cp310
  readline           conda-forge/linux-64::readline-8.1.2-h0f457ee_0
  setuptools         conda-forge/linux-64::setuptools-65.0.2-py310hff52083_0
  sqlite             conda-forge/linux-64::sqlite-3.39.2-h4ff8645_1
  tk                 conda-forge/linux-64::tk-8.6.12-h27826a3_0
  tzdata             conda-forge/noarch::tzdata-2022c-h191b570_0
  wheel              conda-forge/noarch::wheel-0.37.1-pyhd8ed1ab_0
  xopen              conda-forge/linux-64::xopen-1.6.0-py310hff52083_0
  xz                 conda-forge/linux-64::xz-5.2.6-h166bdaf_0
  zlib               conda-forge/linux-64::zlib-1.2.12-h166bdaf_2



Downloading and Extracting Packages
sqlite-3.39.2        | 788 KB    | ################################################################################################################################################################### | 100% 
xz-5.2.6             | 409 KB    | ################################################################################################################################################################### | 100% 
cutadapt-4.1         | 211 KB    | ################################################################################################################################################################### | 100% 
xopen-1.6.0          | 27 KB     | ################################################################################################################################################################### | 100% 
libsqlite-3.39.2     | 789 KB    | ################################################################################################################################################################### | 100% 
dnaio-0.9.1          | 80 KB     | ################################################################################################################################################################### | 100% 
python-isal-1.0.1    | 47 KB     | ################################################################################################################################################################### | 100% 
setuptools-65.0.2    | 1.4 MB    | ################################################################################################################################################################### | 100% 
pip-22.2.2           | 1.5 MB    | ################################################################################################################################################################### | 100% 
tzdata-2022c         | 119 KB    | ################################################################################################################################################################### | 100% 
openssl-3.0.5        | 2.8 MB    | ################################################################################################################################################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate cutadapt-4.1
#
# To deactivate an active environment, use
#
#     $ conda deactivate</code></pre>
</div>
</div>
</div>
</div>
</div>
<p><br></p>
</section>
</section>
<section id="activating-conda-environments" class="level3" data-number="5.4">
<h3 data-number="5.4" class="anchored" data-anchor-id="activating-conda-environments"><span class="header-section-number">5.4</span> Activating conda environments</h3>
<p>Whereas we use the term “load” for <code>Lmod</code> modules, we use <strong>“activate”</strong> for conda environments — it means the same thing. Oddly enough, the most foolproof way to activate a conda environment is to use <code>source activate</code> rather than the expected <code>conda activate</code> — for instance:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate cutadapt-4.1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="bash-out">
<p>(cutadapt-4.1) [jelmer@pitzer-login03 PAS2250]$</p>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Environment indicator
</div>
</div>
<div class="callout-body-container callout-body">
<p>When we have an active conda environment, its name is conveniently displayed in our prompt, as depicted above.</p>
</div>
</div>
<p>After we have activated the <code>cutadapt</code> environment, we should be able to actually use the program. To test this, we’ll again simply run it with a <code>--help</code> option:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">cutadapt</span> <span class="at">--help</span> <span class="kw">|</span> <span class="fu">head</span>     <span class="co"># I'm piping into head to avoid pages worth of output </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="bash-out">
<p>cutadapt version 4.1</p>
<p>Copyright (C) 2010-2022 Marcel Martin <a href="mailto:marcel.martin@scilifelab.se" class="email">marcel.martin@scilifelab.se</a></p>
<p>cutadapt removes adapter sequences from high-throughput sequencing reads.</p>
<p>Usage:<br>
&nbsp; &nbsp; &nbsp; &nbsp; cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq</p>
<p>For paired-end reads:</p>
</div>
<p><br></p>
</section>
<section id="creating-an-environment-for-any-program" class="level3" data-number="5.5">
<h3 data-number="5.5" class="anchored" data-anchor-id="creating-an-environment-for-any-program"><span class="header-section-number">5.5</span> Creating an environment for <em>any</em> program</h3>
<p>Minor variations on the <code>conda create</code> command above can be used to install almost any program for which is conda package is available. However, you may be wondering how we would know:</p>
<ul>
<li>Whether the program is available and what its conda package’s name is</li>
<li>Which conda channel we should use</li>
<li>Which versions are available</li>
</ul>
<p>My strategy to finding these things out is to simply Google the program name together with “conda”, e.g.&nbsp;<code>cutadapt conda</code>.</p>
<p>Let’s see that in action:</p>
<p align="center">
<img src="img/conda_google.png" width="85%">
</p>
<p><br></p>
<p>We click on that first link (it should always be the first Google hit):</p>
<p align="center">
<img src="img/conda_website.png" width="85%">
</p>
<p>I always take the top of the two example installation commands as a template, here: <code>conda install -c bioconda cutadapt</code></p>
<p>You may notice the <code>install</code> subcommand, which we haven’t yet seen. This would install Cutadapt into <em>the currently activated conda environment</em>. Since our strategy here –and my general strategy– is to create a new environment each time you’re installing a program, the all-in-one command is to replace <code>install</code> by <code>create -y -n &lt;env-name&gt;</code>.</p>
<p>Then, our full command (without version specification) again becomes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-y</span> <span class="at">-n</span> cutadapt <span class="at">-c</span> bioconda cutadapt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To see which <strong>version</strong> will be installed by default, and to see which older versions are available:</p>
<p align="center">
<img src="img/conda_website_version.png" width="70%">
</p>
<p>For almost any other program, this works exactly the same!</p>
</section>
<section id="lines-to-add-to-your-bash-script" class="level3" data-number="5.6">
<h3 data-number="5.6" class="anchored" data-anchor-id="lines-to-add-to-your-bash-script"><span class="header-section-number">5.6</span> Lines to add to your Bash script</h3>
<p>While you’ll typically want to do installation interactively and only need to do to it once (see note below), you should always include the necessary code to load/activate your programs in your shell scripts.</p>
<p>When your program is in an <code>Lmod</code> module, this simply entails a <code>module load</code> call — e.g., for <code>fastqc</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="bu">set</span> <span class="at">-ueo</span> pipefail</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Load software</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load fastqc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>When your program is available in a conda environment, this entails a <code>module load</code> command to load conda itself, followed by a <code>source activate</code> command to load the relevant conda environment:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Load software</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load miniconda3</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate cutadapt-4.1</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Strict/safe Bash settings </span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a><span class="bu">set</span> <span class="at">-ueo</span> pipefail</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>We’ve moved the <code>set -ueo pipefail</code> line <em>below</em> the <code>source activate</code> command, because the conda activation procedure can sometims throw “unbound variable” errors otherwise.</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Install once, load/activate always
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><p>Provided you don’t need to switch versions, <strong>you only need to install a program once</strong>. This is true also at OSC and also when using conda: your environments won’t disappear unless you delete them.</p></li>
<li><p>In every single “session” that you want to use a program via an <code>Lmod</code> module or a conda environment, you need to load/activate the program. So the line(s) to do so should always be in your script for that program.</p></li>
</ul>
</div>
</div>
<p><br> <br></p>
<hr>
<hr>
</section>
</section>
<section id="addendum-a-few-other-useful-conda-commands" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="addendum-a-few-other-useful-conda-commands"><span class="header-section-number">6</span> Addendum: a few other useful conda commands</h2>
<ul>
<li><p>Deactivate the currently active conda environment:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate   </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></li>
<li><p>Activate one environment and then “stack” an <em>additional</em> environment (a regular <code>conda activate</code> command would <em>switch</em> environments):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate cutadapt         <span class="co"># Now, the env "cutadapt" is active</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate <span class="at">--stack</span> multiqc   <span class="co"># Now, both "cutadapt" and "multiqc" are active</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></li>
<li><p>Remove an environment entirely:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> env remove <span class="at">-n</span> cutadapt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></li>
<li><p>List all your conda environments:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> env list</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></li>
<li><p>List all packages (programs) installed in an environment:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> list <span class="at">-n</span> cutadapt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></li>
</ul>


</section>


<div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>When your personal computer asks you to “authenticate” while you are installing something, you are authenticating yourself as a user with administrator privileges. At OSC (and for OSU-managed personal computers, too!), you don’t have such privileges.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>Other software upon which the software that you are trying to install depends.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>And this is certainly worth considering if you find yourself using conda a lot, because the conda version at OSC is quite out-of-date.<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">2023, Jelmer Poelstra</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/mcic-osu/rnaseq-intro">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>



</body></html>